### PR TITLE
docs(appium): create new docs generator for appium 2

### PR DESCRIPTION
**Pull request by jlipps**
Things this PR does:

### Delete things

* all of the old prose guides/tutorials/docs (in `docs`) that were driver-specific. These should be reincarnated in the driver repos themselves based on what's correct from appium/master.
* all of the old command documentation (yml files, plus the markdown files generated from them). we will be moving to JSDoc for automatically generating command docs based on code; plus, most of the command documentation needs to live with the drivers and not the umbrella server docs.
* all of the old sample code. it was most assuredly wrong. we should rebuild and make sure that these things are included in CI so they don't break.

### Move things

* some of the current prose guides/tutorials/docs from `docs` have been moved to `packages/appium/docs`
* moved the current Japanese docs into `packages/appium/docs/ja`

### Add things

* added `packages/docutils` as a package we can use for building various appium module docs. it contains:
    * a base jsdoc config
    * a base mkdocs config
    * helper methods for running jsdoc and mkdocs
* added `packages/appium/docs` as the new home for (pre-built) appium docs

### Overview of the new docs architecture

Here's what the architecture in PR is going for:

* write docs in `packages/appium/docs/LANG`
* structure docs site via the `mkdocs-LANG.yml` file
* build docs (one for each language) using [mkdocs](https://www.mkdocs.org/) with the [mkdocs-material](https://squidfunk.github.io/mkdocs-material/) theme. this theme has a _lot_ of benefits and is super extensible.
* rely on multiple levels of base mkdocs config (`packages/docutils/base-mkdocs.yml` and `packages/appium/docs/base-mkdocs.yml`) to make config for specific appium doc sets easy
* generate API/command docs via jsdoc
* package all this doc building logic up into a `docutils` module that can be used via npm scripts by any of our modules or by third-party modules to build their own docs sites.
* publish docs such that previous versions of the docs stay associated with their published appium version and continue to be referenced. i.e., versioned docs!
 
### Things to do after this is merged

* Actually write any stubbed out docs (there's a lot here to do!)
* Figure out what should go at https://appium.io (should it be a marketing landing page? should it autodirect to /docs/en?)
* generate API docs for BaseDriver
* generate API docs (if any) for Appium or AppiumDriver
* figure out how to combine the various docs into the mkdocs site
* figure out how to generate/publish all docs via CI
* localization work
    * update the Japanese docs
    * invite writing of updated Chinese docs (someone can either move and create a good TOC, or rewrite from scratch)
* move old platform-specific docs to the relevant driver repos
* implement doc sites for all the official drivers (should be as easy as breaking the README out and using the same site generation techniques as we are using here. we can add command/API docs over time).
 
### How to try out this PR

* Pull down the branch and do a reinstall just in case
* Make sure that `mkdocs` and `mkdocs-material` and `mike` are installed (e.g., `pip3 install mkdocs mkdocs-material mike`)
* `cd packages/appium`
* `npm run dev:docs`
* go to http://localhost:8000 in the browser and start exploring. the quickstart is done for JS, for example.
* or `npm run build:docs`. docs will be built by language and version into the `gh-pages` branch. you probably want to clean up and reset your gh-pages branch if you have one, first.
* `npm run publish:docs` if you want to actually commit the docs to the `gh-pages` branch.

### CODE REVIEW

**Comment by KazuCocoa** [comment]
Took a look. I liked that, nice 👍 

> Figure out what should go at https://appium.io/ (should it be a marketing landing page? should it autodirect to /docs/en?)

I think `docs/en/` can be the top since other pages haven't been maintained well like several images.

**Comment by jlipps** [comment]
OK @boneskull @KazuCocoa @mykola-mokhnach I think this is ready for review. If you want to see the docs that it publishes, I have it publishing to the gh-pages branch which is visible here: https://appium.github.io/appium

**Comment by jlipps** [comment]
fyi it sounds like @boneskull is going to explore a different static site generator in case we can get the same functionality with a JS-based one rather than a Python-based one

**Comment by KazuCocoa** [comment]
oh, cool. I wondered if we could use only JS tools. Then, we can simplify the appium.io tool chain. (currently it uses Python and Ruby)

**Comment by jlipps** [comment]
> Then, we can simplify the appium.io tool chain

With this PR, we will be able to simply move appium.io to the boneyard, since everything is included to push to the gh-pages branch of this repo. That is one thing I am looking forward to! And while this PR does require python, it does not require ruby.

**Comment by jlipps** [comment]
@mykola-mokhnach @KazuCocoa any objections to me merging this so new additions to the docs can come easily reviewable in fresh PRs? i will also work on making a list of docs we need to write out so we can assign volunteers

**Comment by KazuCocoa** [comment]
No. I'm ok to merge this and go ahead to prepare documentations. @jlipps 

Then, we no longer maintain https://github.com/appium/appium/tree/master/docs basically. We can accept small changes by community to the master, and apply the diff to the 2.0 branch as a baseline. Correct?

**Review by KazuCocoa — APPROVED** [review]
it needs to fix the conflict, otherwise lg

**Comment by boneskull** [comment]
For the record:

I'd prefer we use a JS-based SSG, but I spent some time investigating what this would take. Eleventy works with Nunjucks templates, which is a rough port of Jinja2 to JS. To that end I was able to get the templates in mkdocs-material building and running via Eleventy (sans a couple plugins)... but the frontend build process of mkdocs-material is so complicated that maintaining our own Nunjucks fork of the project would be incredibly daunting. 

So, could attempt the above, or build an entire site from scratch or some _other_ boilerplate template, but it's going to be hard to find one that fits our needs--or just bite the bullet and keep mkdocs. 

To be clear: because we're using a python toolchain, any extension author who wants to build docs using our tools will need to be running a python toolchain. However, the dependencies aren't too bad, and it remains to be seen what kind of stumbling block--if any--this will be for the developer community. 

In summary, we should just go with this for now. 

My recommendation is to donate to the Patreon of the author of mkdocs-material. 🙂

**Comment by jlipps** [comment]
> My recommendation is to donate to the Patreon of the author of mkdocs-material

I agree, I think this also unlocks some features which might be useful

**Comment by jlipps** [comment]
ok, conflicts fixed. @mykola-mokhnach just want to get a 👍  from you as well.

**Comment by boneskull** [comment]
@jlipps remove `--require ../../test/setup-babel.js` from the test script(s) in `docutils`

**Review by mykola-mokhnach — APPROVED** [review]

**Comment by steffkes** [comment]
this is the first occasion I'm able to find; there are a _ton_ of references all over the place, linking to `/sample-code` in this repository. with this pull request it's just gone :/

How can I help with this? I'm interested in Appium and just had a look - wanting to explore those samples, but didn't find them anywhere. The [getting started website](https://appium.io/docs/en/about-appium/getting-started/index.html) for examples mentions them twice and also [within the repository it gets reference quite a few times](https://github.com/appium/appium/search?q=sample+code)

**Comment by jlipps** [comment]
Hi @steffkes thanks for wanting to help. The old docs and sample code were mostly erroneous at this point so making them inaccessible is actually doing the users something of a favour. The new docs are all being developed in `packages/appium/docs/en`, which you can have a look at to see what's written so far. There are a few guides and getting started samples. There are also a lot of docs related issues in our Appium 2 project board. Some of them basically involve moving docs from the `1.x` branch and simply pasting / updating into the appropriate place for the new Appium 2 docs. That would be a chore we would much appreciate help with!

**Comment by steffkes** [comment]
Thanks @jlipps! 

One thing which might help without much work: we re-create the `sample-code` directory and put a small `README` in it which explains the situation. this way, new users get an idea where to go and what to do until they situation is cleaned up

After we have this in place, we could approach the migration that you mentioned. I'll try to get a bit more familiar with the entire appium project so that I have a proper idea what means what and what y'all are talking about here =)

WDYT?

**Comment by jlipps** [comment]
sounds great. as your first PR want to create this directory and simple README? you could also point people to the `1.x` branch where all that old stuff still lives.
